### PR TITLE
ASCII fast paths for grapheme stride

### PR DESCRIPTION
**Pull request by Catfish-Man**
Fixes rdar://124028755 (ASCII fast paths for grapheme stride)

### CODE REVIEW

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci Please Apple Silicon benchmark

**Comment by Catfish-Man** [comment]
@swift-ci Please benchmark

**Comment by Catfish-Man** [comment]
@swift-ci Please Apple Silicon benchmark

File: stdlib/public/core/StringGraphemeBreaking.swift [hunk 1]
```diff
@@ -196,9 +196,30 @@ extension _StringGuts {
...
   internal func _opaqueCharacterStride(startingAt i: Int) -> Int {
+    _internalInvariant(i < endIndex._encodedOffset)
+    if isFastUTF8 {
+      let fast = withFastUTF8 { utf8 in
+        if i &+ 1 == utf8.count { return true }
+        let pair = UnsafeRawPointer(
+          utf8.baseAddress.unsafelyUnwrapped
+        ).loadUnaligned(fromByteOffset: i, as: UInt16.self)
+        //& 0x8080 == 0 is "both not ASCII", != 2573 is "not CRLF"
+        return pair & 0x8080 == 0 && pair != 2573
```
Comment (milseman) [inline stdlib/public/core/StringGraphemeBreaking.swift:210]: CF-LF would probably read better in hex, otherwise LGTM

**Review by milseman — APPROVED** [review]

**Review by milseman — APPROVED** [review]

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
@swift-ci Please Apple Silicon benchmark

**Comment by Catfish-Man** [comment]
x86 benchmark results (-O is similar)

```
------- Performance (x86_64): -Osize -------

REGRESSION                                 OLD         NEW         DELTA    RATIO    
FlattenListFlatMap                         2736.0      3397.0      +24.2%   **0.81x (?)**
String.replaceSubrange.String              9.676       11.216      +15.9%   **0.86x (?)**
StringSwitch                               224.875     258.429     +14.9%   **0.87x (?)**
StringBuilder                              233.125     263.0       +12.8%   **0.89x (?)**
NSStringConversion.InlineBuffer.UTF8       639.0       714.0       +11.7%   **0.89x (?)**
ObjectiveCBridgeStringHash                 66.886      73.786      +10.3%   **0.91x (?)**
StringBuilderSmallReservingCapacity        241.857     265.714     +9.9%    **0.91x (?)**
String.replaceSubrange.String.Small        38.714      42.333      +9.3%    **0.91x (?)**
BridgeString.find.native.longNonASCII      419.75      457.0       +8.9%    **0.92x (?)**
NormalizedIterator_latin1                  182.87      197.905     +8.2%    **0.92x (?)**

IMPROVEMENT                                OLD         NEW         DELTA    RATIO    
StringDistance.characters.ascii            2887.0      975.5       -66.2%   **2.96x**
FlattenDistanceFromTo.Array.String.08.08   2208.0      901.5       -59.2%   **2.45x**
FlattenDistanceFromTo.Array.String.04.08   323.5       150.0       -53.6%   **2.16x**
FlattenDistanceFromTo.Array.String.08.04   357.167     192.636     -46.1%   **1.85x**
FlattenDistanceFromTo.Array.String.04.04   55.725      31.609      -43.3%   **1.76x**
StringHasPrefixAscii                       3611.667    2162.5      -40.1%   **1.67x**
StringHasSuffixAscii                       3506.667    2315.0      -34.0%   **1.51x**
FindString.Loop1.Substring                 327.143     251.0       -23.3%   **1.30x**
LessSubstringSubstring                     31.897      24.923      -21.9%   **1.28x**
RomanNumbers2                              451.5       353.0       -21.8%   **1.28x (?)**
StringHasPrefixUnicode                     40037.037   31379.31    -21.6%   **1.28x (?)**
EqualSubstringSubstring                    30.267      23.844      -21.2%   **1.27x (?)**
EqualSubstringSubstringGenericEquatable    30.269      23.85       -21.2%   **1.27x**
EqualStringSubstring                       30.25       23.842      -21.2%   **1.27x (?)**
EqualSubstringString                       31.25       24.829      -20.5%   **1.26x (?)**
LessSubstringSubstringGenericComparable    31.031      25.031      -19.3%   **1.24x**
RemoveWhereQuadraticString                 291.6       237.0       -18.7%   **1.23x**
UTF8Decode_InitFromBytes                   179.0       145.917     -18.5%   **1.23x (?)**
UTF8Decode_InitFromData                    173.583     142.308     -18.0%   **1.22x**
UTF8Decode_InitDecoding                    170.6       142.0       -16.8%   **1.20x (?)**
UTF8Decode_InitFromCustom_contiguous       171.0       142.615     -16.6%   **1.20x (?)**
FindString.Rec3.Substring                  136.933     115.176     -15.9%   **1.19x (?)**
RemoveWhereFilterString                    223.545     189.833     -15.1%   **1.18x**
StringMatch                                5665.217    4820.0      -14.9%   **1.18x**
StringComparison_longSharedPrefix          239.778     205.0       -14.5%   **1.17x (?)**
StringHasSuffixUnicode                     64000.0     55043.478   -14.0%   **1.16x (?)**
FindString.Rec3.String                     141.308     122.438     -13.4%   **1.15x (?)**
SubstringTrimmingASCIIWhitespace           143.0       124.929     -12.6%   **1.14x (?)**
LineSink.characters.alpha                  77.774      68.171      -12.3%   **1.14x (?)**
StringRemoveDupes                          240.875     213.889     -11.2%   **1.13x**
CSVParsing.Char                            163.643     148.938     -9.0%    **1.10x (?)**
UTF8Decode_InitFromCustom_noncontiguous    331.167     304.286     -8.1%    **1.09x (?)**
CStringLongNonAscii                        156.5       143.917     -8.0%    **1.09x (?)**
FrequenciesUsingReduceInto                 711.538     657.692     -7.6%    **1.08x (?)**
LineSink.characters.complex                376.0       350.0       -6.9%    **1.07x (?)**
```

**Comment by Catfish-Man** [comment]
ARM benchmark results match what I saw locally, so I think this is good to go if tests look fine :) 

```
------- Performance (arm64): -Osize -------

REGRESSION                                 OLD        NEW         DELTA    RATIO    
CharIteration_russian_unicodeScalars       3153.103   3652.8      +15.8%   **0.86x**
StringComparison_nonBMPSlowestPrenormal    338.358    377.581     +11.6%   **0.90x**
StringComparison_emoji                     185.49     203.478     +9.7%    **0.91x**

IMPROVEMENT                                OLD        NEW         DELTA    RATIO    
StringDistance.characters.ascii            1760.0     507.75      -71.2%   **3.47x**
FlattenDistanceFromTo.Array.String.08.08   1325.0     571.25      -56.9%   **2.32x**
FlattenDistanceFromTo.Array.String.04.08   196.083    92.083      -53.0%   **2.13x**
StringHasSuffixAscii                       2278.0     1310.556    -42.5%   **1.74x**
FlattenDistanceFromTo.Array.String.08.04   214.727    123.889     -42.3%   **1.73x**
FlattenDistanceFromTo.Array.String.04.04   33.484     20.033      -40.2%   **1.67x**
StringHasPrefixAscii                       2402.222   1529.333    -36.3%   **1.57x**
StringHasSuffixUnicode                     42625.0    31564.516   -25.9%   **1.35x**
RemoveWhereQuadraticString                 163.154    124.118     -23.9%   **1.31x**
FindString.Loop1.Substring                 184.692    144.588     -21.7%   **1.28x**
RomanNumbers2                              250.125    196.364     -21.5%   **1.27x**
RemoveWhereFilterString                    116.667    99.48       -14.7%   **1.17x**
StringMatch                                3286.441   2802.899    -14.7%   **1.17x**
StringUTF16SubstringBuilder                910.0      792.857     -12.9%   **1.15x**
FindString.Rec3.Substring                  67.559     59.1        -12.5%   **1.14x**
FindString.Rec3.String                     66.882     59.256      -11.4%   **1.13x (?)**
StringRemoveDupes                          154.286    141.313     -8.4%    **1.09x**
LineSink.characters.alpha                  51.417     47.231      -8.1%    **1.09x**
SubstringTrimmingASCIIWhitespace           83.556     77.828      -6.9%    **1.07x**
Breadcrumbs.MutatedIdxToUTF16.ASCII        1.395      1.3         -6.8%    **1.07x**
Breadcrumbs.MutatedUTF16ToIdx.ASCII        1.396      1.301       -6.8%    **1.07x**
FrequenciesUsingReduceInto                 445.769    415.517     -6.8%    **1.07x (?)**
```

**Comment by Catfish-Man** [comment]
@swift-ci please test windows platform

**Pull request merged**